tests/test_templatetags.py

```diff
@@ -0,0 +1,122 @@
+
+from djangocms_snippet.models import Snippet, SnippetPtr
+
+
+class SnippetTemplateTagTestCase(TestCase):
+
+    def test_html_rendered(self):
+        snippet = Snippet.objects.create(
+            name="test snippet",
+            html="<p>hello {{ title }}</p>",
+            slug="test_snippet",
+        )
+        SnippetPtr.objects.create(
+            snippet=snippet,
+        )
+
+        context = Context({"title": "world"})
+        template_to_render = Template(
+            '{% load snippet_tags %}'
+            '{% snippet_fragment "test_snippet" %}'
+        )
+        rendered_template = template_to_render.render(context)
+        self.assertInHTML('<p>hello world</p>', rendered_template)
+
+        # test html errors
+        context = Context({"title": "world"})
+        template_to_render = Template(
+            '{% load snippet_tags %}'
+            '{% snippet_fragment "test_snippet_2" %}'
+        )
+        with self.assertRaises(ObjectDoesNotExist):
+            # Snippet matching query does not exist.
+            rendered_template = template_to_render.render(context)
+
+    def test_template_rendered(self):
+        template = "snippet.html"
+        snippet = Snippet.objects.create(
+            name="test snippet",
+            template=template,
+            slug="test_snippet",
+        )
+        SnippetPtr.objects.create(
+            snippet=snippet,
+        )
+
+        # use a string to identify
+        context = Context({})
+        template_to_render = Template(
+            '{% load snippet_tags %}'
+            '{% snippet_fragment "test_snippet" %}'
+        )
+        rendered_template = template_to_render.render(context)
+        self.assertInHTML('<p>Hello World Template</p>', rendered_template)
+
+        # use an id to identify
+        context = Context({})
+        template_to_render = Template(
+            '{% load snippet_tags %}'
+            '{% snippet_fragment 1 %}'
+        )
+        rendered_template = template_to_render.render(context)
+        self.assertInHTML('<p>Hello World Template</p>', rendered_template)
+
+        # tests "or" functionality
+        context = Context({})
+        template_to_render = Template(
+            '{% load snippet_tags %}'
+            '{% snippet_fragment "test_snippet_1" or %}<p>hello world</p>{% endsnippet_fragment %}'
+        )
+        rendered_template = template_to_render.render(context)
+        self.assertInHTML('<p>hello world</p>', rendered_template)
+
+        # failure when context is set incorrectly
+        context = ""
+        template_to_render = Template(
+            '{% load snippet_tags %}'
+            '{% snippet_fragment "test_snippet_1" or %}<p>hello world</p>{% endsnippet_fragment %}'
+        )
+        with self.assertRaises(AttributeError):
+            # AttributeError: 'str' object has no attribute 'render_context'
+            rendered_template = template_to_render.render(context)
```

Comment:
This last case seems unnecessary to me – I mean, this violates the most basic API of Django templates, which expects `render` to be called with a `Context` instance; ideally, it would blow up with a `TypeError`, but the fact that it raises an `AttributeError` is just an implementation detail. Anyway, I see no point in testing this particular invalid case in our own code.

Reply:
👍